lib/Sema/CSSimplify.cpp

```diff
@@ -3939,9 +3940,19 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
     break;
   }
   
+  // FIXME: Allow the type to be upcast if the type system has a concrete
+  // KeyPath type assigned to the expression already.
+  if (keyPathBGT) {
+    if (keyPathBGT->getDecl() == getASTContext().getKeyPathDecl())
+      kpDecl = getASTContext().getKeyPathDecl();
+    else if (keyPathBGT->getDecl() == getASTContext().getWritableKeyPathDecl()
+             && capability >= Writable)
+      kpDecl = getASTContext().getWritableKeyPathDecl();
+  }
```

Comment:
I just pulled it up in the debugger. I see a couple issues:
- The handling for `ConstraintKind::KeyPath` and `ConstraintKind::KeyPathApplication` in `getPotentialBindings()` is wrong. I believe it should be setting `InvolvesTypeVariables` if there are unbound type variables (other than the one we're getting bindings for) in the constraints involving this type variable.

So for one failing test case we have:
```
  $T5 key path from $T1 -> $T4
  $T5 conv KeyPath<A, A>
```
and the code as is would let us select bindings for `$T5` even if we hadn't bound `$T1` and `$T4`.

- The other problem is that even if those *are* bound we only consider `KeyPath<A,A>` as a binding for `$T5` (as a result of the conversion constraint), and we'll attempt that binding before trying to process the constraints.

So I *think* what should be happening here is that rather than binding the type variable during constraint processing you should be adding the type you would bind it to as a potential binding in `getPotentialBindings()`. It's not clear to me from a quick glance if you have enough information at that point to do so. Then during constraint processing you would verify that the types involved satisfy the constraint.

Although I've noticed other code calling `matchTypes` directly with `ConstraintKind::Bind` during constraint simplification, this is really not ideal because you have no control over the order of processing constraints and/or the relative order with respect to other type variable bindings.